Ensure that step phases have unique names

therazix opened the pull request:
If discover step has multiple phases with the same name, tmt will raise an error with the corresponding message.

Fix: #2029 

Should I also add a check to `tmt lint`?

happz commented:
Nice, I missed the bug report, thanks for looking into it.

In general, *all* steps shall require their phases to have different names, not just the `discover`. E.g. I wouldn't trust tmt with two guests with the same `name` key. Addressing this would require a change of your patch:

* https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L276 is where a step implementation generates names for phases that do not have one,
* https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L451 is where a step implementation ingests CLI input, and
* https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L286 is where a step implementation converts a `dict`-like structures - like fmf nodes or CLI options - into dataclasses tmt uses for internal representation of step data.

So, these are places where `name` keys of phases enter tmt machinery. My proposal would therefore be to 1. extract the name generation into a stand-alone function, so it could be called from the place where it's now and from the CLI ingestion spot as well (that would be new), and 2. add your test as the second stand-alone function. I'd like to keep them separated, for easier unit testing. We would then call them from https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L275 and https://github.com/teemtee/tmt/blob/main/tmt/steps/__init__.py#L489, when the step is done with ingesting data from fmf or CLI, respectively.

It would be better to put these two new calls to `_normalize_data()`, but I worried that it might delay the error, by having them placed after here could provide user with an earlier error.

Plus, yep, `tmt lint` checks would be awesome, and you could build it from these two new functions: load fmf data (already done by tmt), add missing `name` keys (1st new function), check there are no duplicities (2nd new function) & if there were duplicities, an exception is turned into `LinterOutcome` :)

therazix commented:
Thank you for the feedback, I updated the PR. The error is now raised if any step has multiple phases with the same `name`. I also added the lint check.

I put the name-checking call inside the `_normalize_data()` because otherwise the error will be thrown even if only `tmt lint` is executed.

If I should make any changes, let me know.

happz commented:
@therazix LGTM, thanks for the changes. With your patch, I believe we don't need a custom method in the `provision` step anymore, https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/__init__.py#L1519. It's also related to the failing tests, so after this is resolved, the PR could be on its way to `main`.

happz commented on tmt/base.py:
```diff
@@ -1957,6 +1957,23 @@ def lint_fmf_remote_ids_valid(self) -> LinterReturn:
+                    else:
+                        seen.append(name)
+        if passed:
+            yield LinterOutcome.PASS, 'phases have unique names'
```
One more simplification maybe: `lint_uniq_names` and `_check_duplicate_names` do share some functionality, looking for duplicates. How about extracting that into a helper function? We already have a useful `tmt.utils.uniq`, I would add `tmt.utils.duplicates` to complement it:

```python
def duplicates(values: Iterable[Optional[T]]) -> Generator[T, None, None]:
    seen = collections.Counter(values)

    for value, count in seen.items():
        if value is None or count == 1:
            continue

        yield value

# And then in base.py...
def lint_unique_names(self) -> LinterReturn:
    """ P006: phases must have unique names """
    passed = True
    for step_name in self.step_names(enabled=True, disabled=True):
        for phase_name in duplicates(phase.get('name', None) for phase in self._step_phase_nodes(step_name)):
            passed = False
            yield LinterOutcome.FAIL, \
                f'duplicate phase name "{name}" in step "{step_name}"'

    if passed:
        yield LinterOutcome.PASS, 'phases have unique names'

# ... and steps/__init__.py:
def _check_duplicate_names(self, raw_data: List[_RawStepData]) -> None:
    """ Check for duplicate names in phases """

    for name in duplicates(raw_datum.get('name', None) for raw_datum in raw_data):
        raise tmt.utils.GeneralError(f"Duplicate phase name '{name}' in step '{self.name}'.")
```

We get a friend for `uniq`, a nicely isolated building block, a bit of a functional twist plus more readable methods.

therazix replied:
Thanks for your feedback. I extracted the shared functionality into a separate function, as you suggested. I also added a unit test for this new function.

happz replied:
Nice!

psss reviewed (approved):
Nice! Thanks much for fixing this. Added just a trivial adjustment in 65bb2223.
